### ISSUE TITLE
poor diagnostic due to overloading in `extern "C"` block

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#80235]
https://godbolt.org/z/Koo18frzd (GCC does a bit of a better job, treating them as redeclarations and mentioning they're C functions)

The following invalid C++ code:
```
extern "C" {
void Foo(int);
int Foo();
}
```
will give the error:
```
<source>:5:5: error: functions that differ only in their return type cannot be overloaded
    5 | int Foo();
      | ~~~ ^
<source>:4:6: note: previous declaration is here
    4 | void Foo(int);
      | ~~~~ ^
1 error generated.
```
which is profoundly unfair to the person writing the code. The error isn't anything to do with functions that differ only in their return type, and everything to do with being in an extern "C" block, where overloading does not exist.

(quoting @ssbr from an internal bug)

**Comment by shafik** [comment]
Confirmed.

The diagnostic is here:

https://github.com/llvm/llvm-project/blob/0c361270afff83cd6433cf865ed5a410dadfb33f/clang/lib/Sema/SemaDecl.cpp#L3945-L3948

if we modified the `if` statement:

https://github.com/llvm/llvm-project/blob/0c361270afff83cd6433cf865ed5a410dadfb33f/clang/lib/Sema/SemaDecl.cpp#L3921

to

```cpp
if (getLangOpts().CPlusPlus && !(New->isExternC() && Old->isExternC())) {
```

The diagnostic is:

```console
externC.cpp:3:5: error: conflicting types for 'Foo'
    3 | int Foo();
      |     ^
externC.cpp:2:6: note: previous declaration is here
    2 | void Foo(int);
      |      ^
1 error generated.
```

Which is a lot closer to gcc.

CC @AaronBallman

**Comment by s-watanabe314** [comment]
Is anyone working on this issue now?
I would like to work on it as my first contribution to LLVM.

**Comment by s-watanabe314** [comment]
Thanks for the confirmation. I will give it a try.
I didn't know there was an `overloadable` attribute in clang, so I'll check it out as I go along.

In gcc, the compile error message for the program is as follows:
```
test.cpp:3:13: error: conflicting declaration of C function 'int Foo()'
    3 |         int Foo();
      |             ^~~
test.cpp:2:14: note: previous declaration 'void Foo(int)'
    2 |         void Foo(int);
      |              ^~~
```

This message explicitly states that there is a declaration conflict in C, but there seems to be no such error message in clang.
In order to output the same error message as gcc, I think we need to implement a new error message and branching process, is the above suggested modification to output 
`externC.cpp:3:5: error: conflicting types for 'Foo'`
sufficient?

I would appreciate any advice anyone can give me.

**Comment by dwblaikie** [comment]
I think the first/lower-cost fix would be /just/ to avoid the "differ only in their return type" misleading message, see this example: https://godbolt.org/z/Wj3xc1h96
```
<source>:3:6: error: conflicting types for 'f1'
    3 | void f1(int);
      |      ^
<source>:2:6: note: previous declaration is here
    2 | void f1(void);
      |      ^
<source>:5:5: error: functions that differ only in their return type cannot be overloaded
    5 | int f2(int);
      | ~~~ ^
<source>:4:6: note: previous declaration is here
    4 | void f2(void);
      | ~~~~ ^
```
If you can make the second case look more like/the same as the first, that'd be great.

Improving the /first/ case, then (once it includes the second case) to look more like GCC I think would be a separate issue/bug fix - which you're welcome to contribute, separately.

**Comment by s-watanabe314** [comment]
Thanks for the reply.  
This time I will modify the code so that the second case uses the same error statement as the first case.  
I will check to see if there is a separate GitHub issue related to outputting error messages like GCC. If not, I will consider posting an issue.

**Comment by s-watanabe314** [comment]
The suggested patch didn't work well. I have found that other regression tests fail.
So I tried another modification.

I'm modifying the code at
 https://github.com/llvm/llvm-project/blob/0c361270afff83cd6433cf865ed5a410dadfb33f/clang/lib/Sema/SemaDecl.cpp#L3941-L3948
 by inserting the following two lines before line 3944:


        else if (Old->isExternC() && New->isExternC())
          Diag(New->getLocation(), diag::err_conflicting_types) << New;

After this modification, the expected diagnostic for this issue is correctly output. However, the diagnostic for merging functions with the `overloadable` attribute has changed.


Example:

```
extern "C" {
  __attribute__((overloadable))
  void ovl_decl(void);

  __attribute__((overloadable))
  int ovl_decl(void); // expected-error
}
```
Before the modification:


The diagnostic was: `"functions that differ only in their return type cannot be overloaded"`


After the modification:


The diagnostic is: `"conflicting types for 'ovl_decl'"`


Is the `"conflicting types ..."` diagnostic appropriate?


If Clang tries to merge the `overloadable` attribute in a `.c` source file, the diagnostic would be `"conflicting types ..."` as seen in https://godbolt.org/z/nGxvde78j. Therefore, I don't see a problem with the new diagnostic.
